content/post/quic/article.md

```diff
@@ -0,0 +1,80 @@
+
+```bash
+ipfs config --json Experimental.QUIC true
+```
+
+Then edit the `~/.ipfs/config` to add a QUIC swarm addresses (e.g. `/ip4/0.0.0.0/udp/4001/quic`, if you want QUIC to listen on UDP port 4001).
+
+## Introduction
+
+QUIC is a new UDP based internet transport that was invented by Google in 2013 and rolled out as an experiment in Google Chrome and Google's frontend servers over the next few years. According to Google's measurements, on lossy connections QUIC reduces search latency (the time it takes to fully load the Google Search page) by 3.6 to 8% and the Youtube rebuffer rate (the number of times a YouTube video stops to load new data) by 15 to 18%.
+
+In 2016, the IETF formed a Working Group to specify the QUIC protocol, which is scheduled to complete the protocol specification by mid 2019.
+
+![](stack.png)
+
+QUIC replaces a big part of the current protocol stack of the web:
+
+- it runs on top of UDP, so it has to take care of loss recovery
+- it runs a cryptographic handshake and takes care of encrypting the connection
+- it provides streams at the transport level
+
+In this blog post, I will show how QUIC solves multiple problems inherent to the traditional network stack.
+
+## Handshake Latency
+
+![tcp-handshake](tcp-handshake.png)
+
+In order to establish a new encrypted connection between two hosts (the same applies to TLS 1.2 connection), the client first initiates a TCP connection. The three-way handshake necessary to establish a TCP connection takes one network roundtrip time. After that, the client initiates the cryptographic handshake (secio for libp2p connections, or TLS for browsers). Establishing a secio or TLS 1.2 connection takes another two roundtrip times. Once that is done, client and server are set up to exchange application data.
+
+![](quic-handshake.png)
+
+By using QUIC we can reduce the handshake latency down to a single roundtrip. Since QUIC runs on top of UDP, we don't need the three-way handshake. By using TLS 1.3 instead of secio or TLS 1.2, we save another roundtrip. With TLS 1.3, it's even possible to eliminate handshake latency altogether by doing a so called 0-RTT handshake. The only requirement for this is that the peers already established a connection before, and exchanged some key material.
+
+## Stream Multiplexing
+
+![](http2.png)
+
+libp2p uses streams to transmit data between peers. Streams are a concept that was introduced in Chrome and the Google's frontend servers with the SPDY protocol (which later became HTTP/2). A stream is a birection flow of bytes within an established connection. The advantage of using streams instead of separate connections is that we only need to establish a single TCP connection, and then multiplex a large number of streams on that connection. Opening new streams therefore becomes very cheap. In libp2p, we're using different stream multiplexers, such as yamux and mplex, that are conceptionally very similar to SPDY.
```

Comment:
Add reference links for  yamux and mplex?
```suggestion
libp2p uses streams to transmit data between peers. Streams are a concept that was introduced in Chrome and the Google's frontend servers with the SPDY protocol (which later became HTTP/2). A stream is a birection flow of bytes within an established connection. The advantage of using streams instead of separate connections is that we only need to establish a single TCP connection, and then multiplex a large number of streams on that connection. Opening new streams therefore becomes very cheap. In libp2p, we're using different stream multiplexers, such as [yamux](https://github.com/hashicorp/yamux) and [mplex](https://github.com/libp2p/specs/tree/master/mplex), that are conceptionally very similar to SPDY.
```